dartagnan/src/main/java/com/dat3m/dartagnan/wmm/Wmm.java

```diff
@@ -105,73 +116,116 @@ public void initialise(VerificationTask task, SolverContext ctx) {
         }
     }
 
-    // This methods initializes all relations and encodes all base relations
-    // and recursive groups (why recursive groups?)
-    // It also triggers the computation of may and active sets!
-    // It does NOT encode the axioms nor any non-base relation yet!
-    private BooleanFormula encodeBase(SolverContext ctx) {
-        if (this.task == null) {
-            throw new IllegalStateException("The WMM needs to get initialised first.");
-        }
-
-        for (RecursiveGroup recursiveGroup : recursiveGroups) {
-            recursiveGroup.initMaxTupleSets();
-            recursiveGroup.initMinTupleSets();
-        }
-
-        for (Axiom ax : axioms) {
-            ax.getRelation().getMaxTupleSet();
-        }
-
-        for(String relName : baseRelations){
-            relationRepository.getRelation(relName).getMaxTupleSet();
-        }
-
-        for (Axiom ax : axioms) {
-            ax.getRelation().addEncodeTupleSet(ax.getEncodeTupleSet());
-        }
-
-        Collections.reverse(recursiveGroups);
-        for(RecursiveGroup recursiveGroup : recursiveGroups){
-            recursiveGroup.updateEncodeTupleSets();
-        }
-
-        BooleanFormulaManager bmgr = ctx.getFormulaManager().getBooleanFormulaManager();
+	/**
+	 * Translates the relations of this model into an SMT formula.
+	 * This always includes Read-From, Memory-Order, and Dependencies.
+	 * Computes the may set.
+	 * Computes the must and must-not set.
+	 * Computes the active set.
+	 * @param ctx
+	 * Builder of expressions.
+	 * @return
+	 * Models executions roughly of the associated program.
+	 * Misses control-flow information and parts of intra-thread data-flow information.
+	 * @throws IllegalStateException
+	 * This model has to be invoked with {@link #initialise(VerificationTask, SolverContext)} first.
+	 */
+	public BooleanFormula encodeRelations(SolverContext ctx) {
+		if(task == null) {
+			throw new IllegalStateException("The WMM needs to get initialised first.");
+		}
+
+		//fixed point of may set
+		for(RecursiveGroup recursiveGroup : recursiveGroups) {
+			recursiveGroup.initMaxTupleSets();
+		}
+		for(Axiom ax : axioms) {
+			ax.getRelation().getMaxTupleSet();
+		}
+
+		//fixed point of must and must-not set
+		for(boolean changed = true; changed;) {
```

Comment:
Yes, its like a `while(changed)` but I'm not too big of a fan of this. It saves a line of code but it is quite a bit more obscure. I think it is easier to parse code if `for` is used for enumeration-like purposes and `while` is used for e.g. fixed point computations like this.

Reply:
I agree that it is quite obscure, the simple fact that I have to ask about it is a proof. Please change by `while`.

Reply:
This construction also limits the lifetime of `changed`, so that a similarly-named variable could be used later in the method.  I would have outlined it into a private method, if it occurred more than once.